src/SMPCoroutines.java

```diff
@@ -0,0 +1,217 @@
+            }
+        }
+
+        private void proposeTo(int womanIndex) {
+            Woman next = womenList[womanIndex];
+            System.out.printf("Man %d proposed to %d\n", id, womanIndex);
+            next.receiveProposal(this.id);
+            try {
+                System.out.printf("Man %d waiting for response.\n", id);
+                synchronized (this) {
+                    this.wait();
+                }
+                System.out.printf("Man %d was rejected :(\n", id);
+            } catch (InterruptedException e) {
+
+            }
+        }
+
+    }
+
+
+    class Woman implements Runnable {
+        private int id;
+        private int[] preferences;
+
+        private int acceptedMan;
+        private Man[] menList;
+
+        private Semaphore sem;
+        private int proposedMan;
+
+        public Woman(int id, int[] preferences) {
+            this.id = id;
+            this.preferences = preferences;
+            acceptedMan = -1;
+            sem = new Semaphore(0);
+        }
+
+        public void setMenList(Man[] menList) {
+            this.menList = menList;
+        }
+
+        public void receiveProposal(int proposingMan) {
+            synchronized (sem) {
+                proposedMan = proposingMan;
+                sem.release();
+            }
+        }
+
+        @Override
+        public void run() {
+            while(true) {
+                try {
+                    sem.acquire();
+                    if (acceptedMan < 0) {
+                        acceptedMan = proposedMan;
+                        System.out.printf("Woman %d accepted man %d\n", id, acceptedMan);
+                    } else {
+                        int rankAccepted = -1;
+                        int rankProposed = -1;
+                        for (int i = 0; i < preferences.length; i++) {
+                            int manIndex = preferences[i] - 1;
+                            if (manIndex == proposedMan) {
+                                rankProposed = i;
+                            }
+
+                            if (manIndex == acceptedMan) {
+                                rankAccepted = i;
+                            }
+                        }
+
+                        int manToWakeUp;
+                        if (rankProposed < rankAccepted) {
+                            manToWakeUp = acceptedMan;
+                            acceptedMan = proposedMan;
+                            System.out.printf("Woman %d accepted man %d, ditching man %d\n", id, proposedMan, manToWakeUp);
+                        } else {
+                            manToWakeUp = proposedMan;
+                            System.out.printf("Woman %d rejected man %d, she is already with man %d\n", id, proposedMan, acceptedMan);
+                        }
+
+                        synchronized (menList[manToWakeUp]) {
+                            menList[manToWakeUp].notifyAll();
```

Comment:
If a man is rejected it is time to wake his thread so that he can continue proposing to the rest of his alternatives.